drv/i2c-devices/src/pca9538.rs

```diff
@@ -0,0 +1,163 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Driver for the PCA9538 GPIO expander
+
+use crate::Validate;
+use drv_i2c_api::*;
+use userlib::*;
+
+/// `PinSet` is a bit vector indicating on which pins/ports a given operation is
+/// applied.
+#[derive(Copy, Clone, Debug, PartialEq, FromPrimitive)]
+pub struct PinSet(u8);
+
+impl PinSet {
+    /// Returns a `PinSet` with the mask bit `index` set.
+    #[inline(always)]
+    pub const fn pin(index: usize) -> Self {
+        Self(1 << index)
+    }
+
+    /// Derives a `PinSet` by setting mask bit `index` in addition to the
+    /// already set bits.
+    #[inline(always)]
+    pub const fn and_pin(self, index: usize) -> Self {
+        Self(self.0 | 1 << index)
+    }
+}
+
+/// Derive the union on two `PinSet`s.
+impl core::ops::BitOr for PinSet {
+    type Output = Self;
+
+    fn bitor(self, rhs: Self) -> Self::Output {
+        Self(self.0 | rhs.0)
+    }
+}
+
+/// Pins in a `PinSet` can be configured as either `Input` or `Output`. Note
+/// that even when configured as output, the status of the pin will be reflected
+/// in the result of `read(..)`.
+#[derive(Copy, Clone, Debug, PartialEq, FromPrimitive)]
+#[repr(u8)]
+pub enum Mode {
+    Input = 0,
+    Output = 1,
+}
+
+/// Pins in a `PinSet` can be configured with `Normal` or `Inverted` polarity.
+#[derive(Copy, Clone, Debug, PartialEq, FromPrimitive)]
+#[repr(u8)]
+pub enum Polarity {
+    Normal = 0,
+    Inverted = 1,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, FromPrimitive)]
+enum Register {
+    InputPort = 0x00,
+    OutputPort = 0x01,
+    PolarityInversion = 0x02,
+    Configuration = 0x03,
+}
+
+pub struct Pca9538 {
+    device: I2cDevice,
+}
+
+fn read_reg(
```

Comment:
Nit: `read_reg` and `write_reg` could be member functions on `Pca9538`, then
```rust
let outputs = read_reg(&self.device, Register::OutputPort)?;
```
would become
```rust
let outputs = self.read_reg(Register::OutputPort)?;
```